File: DIP/DIP_007/C++/DIP_007_Otsu's_Binarization.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;

int main()
{
    // Loading the actual image
    Mat image, gray, blurred;
    Mat thresh1, thresh2;

    image=imread("daenerys.jpg", IMREAD_COLOR);
    cv::imshow("Original image", image);
    cv::waitKey();

    // Edge cases
    if(image.empty()){
    	cout << "Error loading image" << endl;
    	return -1;
    }

    // converting the color image into grayscale
    cvtColor(image, gray, COLOR_BGR2GRAY);

    // Otsu's thresholding
    threshold(gray, thresh1, 0, 255, THRESH_BINARY+THRESH_OTSU);

    // Otsu's thresholding with Gaussian filtering
    GaussianBlur(gray, blurred, Size(5,5),0);
    threshold(blurred, thresh2, 0, 255, THRESH_BINARY+THRESH_OTSU);

    // Displaying the result
    cv::imshow("Otsu's thresholding", thresh1);
    cv::imshow("Otsu's thresholding with Gaussian filtering", thresh2);

    cv::waitKey(0);
    return 0;
}

File: DIP/DIP_007/Python/DIP_007_Otsu's_Binarization.py
import cv2

# read the image as a grayscale
image = cv2.imread('daenerys.jpg',0)

# Otsu's thresholding
ret1,th1 = cv2.threshold(image,0,255,cv2.THRESH_BINARY+cv2.THRESH_OTSU)

# Otsu's thresholding with Gaussian filtering
blurred = cv2.GaussianBlur(image,(5,5),0)
ret2,th2 = cv2.threshold(blurred,0,255,cv2.THRESH_BINARY+cv2.THRESH_OTSU)

cv2.imshow("Otsu's thresholding", th1)
cv2.imshow("Otsu's thresholding with Gaussian filtering", th2)
cv2.waitKey(0)
cv2.destroyAllWindows();
